### ISSUE TITLE
A false negative about the detector NN_NAKED_NOTIFY

### ISSUE THREAD

**Issue by jiayiw2008** [spotbugs/spotbugs#3668]
Hi, I found a false negative bug about the rule NN_NAKED_NOTIFY. In the following code example, SpotBugs should have reported a warning at line 7 as the variable `field` is visible to different threads and there are no modifications to mutable objects. This appears to be a false negative.

```java
class NN {
    volatile String field =  NN.class.getCanonicalName();
    public void foo() {
        String str = "str";
        synchronized (field) {
            if (str != null) {
                field.notifyAll(); // should report a warning, but no warnings
            }
        }
    }
}
```

SpotBugs version: 4.9.3

**Comment by JuditKnoll** [comment]
This looks really similar to https://github.com/spotbugs/spotbugs/issues/3634 which you proposed earlier.
@gtoison Does https://github.com/spotbugs/spotbugs/pull/3635 also fix this one too?

**Comment by gtoison** [comment]
I was wondering about that too, I'm not sure and will need to try it out

**Comment by gtoison** [comment]
The `FindNakedNotify` detector is very basic: it gives up when it sees unhandled instructions.
In the above example it resets when it sees `ifnull`.
#3635 does not fix this false negative: we would need a much more substantial change.
